### PR TITLE
Age and Gender Classification using Convolutional Neural Networks

### PR DESCRIPTION
**Pull request by asiryan**
# Age and Gender Classification using Convolutional Neural Networks

## Description
Rothe et al. in their [paper](https://data.vision.ee.ethz.ch/cvl/publications/papers/proceedings/eth_biwi_01229.pdf) propose a deep learning solution to age estimation from a single face image without the use of facial landmarks and introduce the [IMDB-WIKI dataset](https://data.vision.ee.ethz.ch/cvl/rrothe/imdb-wiki/), the largest public dataset of face images with age and gender labels. If the real age estimation research spans over decades, the study of apparent age estimation or the age as perceived by other humans from a face image is a recent endeavor. Rothe et al. tackle both tasks with their convolutional neural networks (CNNs) of VGG-16 architecture which are pre-trained on ImageNet for image classification. They pose the age estimation problem as a deep classification problem followed by a softmax expected value refinement. The key factors of this solution are: deep learned models from large data, robust face alignment, and expected value formulation for age regression. They validate their methods on standard benchmarks and achieve state-of-the-art results for both real and apparent age estimation.

## Model
| Model (Caffe) | Download | ONNX version | Opset version | Dataset |
|:-------------|:--------------|:--------------|:--------------|:--------------|
| [vgg_ilsvrc_16_age_chalearn_iccv2015](https://drive.google.com/drive/folders/1wE4_sj-UBumkjDK9mtfaO9eUan_z44cY?usp=sharing) | [513 MB](https://drive.google.com/file/d/1V75U1kUJ0udBLs6bg3lGqBk3ym8q9guV/view?usp=sharing) | 1.5 | 5 | ChaLearn LAP 2015 |
| [vgg_ilsvrc_16_age_imdb_wiki](https://drive.google.com/drive/folders/14wckle-MbnN10xzdzgF464bMnlM-dd5-?usp=sharing) | [513 MB](https://drive.google.com/file/d/1ECle8EvsXiIid_vMa1_vwMJk6abhzrPF/view?usp=sharing)| 1.5 | 5 | IMDB-WIKI |
| [vgg_ilsvrc_16_gender_imdb_wiki](https://drive.google.com/drive/folders/16Z1r7GEXCsJG_384VsjlNxOFXbxcXrqM?usp=sharing) | [512 MB](https://drive.google.com/file/d/1epLM5ghucLcnGZg-NCIf1r16lotN004I/view?usp=sharing)| 1.5 | 5 | IMDB-WIKI |

## Inference
Input tensor ```1 x 3 x height x width```, which values are in range of ```[0, 255]```. Input image have to be previously resized to ```224 x 224``` pixels and converted to ```BGR``` format.

## Python example
Define downloaded ONNX models and test image in [Python script](age_gender.py)  
```python
age_model = "vgg_ilsvrc_16_age_imdb_wiki.onnx"
gender_model = "vgg_ilsvrc_16_gender_imdb_wiki.onnx"
image_name = "images/brad.jpg"
```
Check the results  
```
Input image: images/brad.jpg
Gender: Man
Age: 41.8
```

## References
* Levi et al. - [Age and Gender Classification Using Convolutional Neural Networks](https://talhassner.github.io/home/publication/2015_CVPR) (**NOT compatible with ONNX**).
* Rothe et al. - [IMDB-WIKI – 500k+ face images with age and gender labels](https://data.vision.ee.ethz.ch/cvl/rrothe/imdb-wiki/).
* Caffe to ONNX: [unofficial converter](https://github.com/asiryan/caffe-onnx).

## Contributors
Asiryan Valery ([asiryan](https://github.com/asiryan))

## License
Apache 2.0

### CODE REVIEW

**Review by wenbingl — COMMENTED** [review]
clean up
vision/body_analysis/age_gender/model/.gitingore 

and why is this file updated?
 vision/classification/mnist/model/mnist-1.tar.gz

**Comment by asiryan** [comment]
> clean up
> vision/body_analysis/age_gender/model/.gitingore
> 
> and why is this file updated?
> vision/classification/mnist/model/mnist-1.tar.gz

Thank you very much for your reply!
1) I cleaned up "models" path with ".gitignore".
2) Because I used recommendation to force push in master to solve DCO exception. It shouldn't be a problem.

**Comment by wenbingl** [comment]
I didn't find the onnx files in your PR, is there something I am missing?
age_model = "**vgg_ilsvrc_16_age_imdb_wiki.onnx**"
gender_model = "**vgg_ilsvrc_16_gender_imdb_wiki.onnx**"

**Comment by asiryan** [comment]
> I didn't find the onnx files in your PR, is there something I am missing?
> age_model = "**vgg_ilsvrc_16_age_imdb_wiki.onnx**"
> gender_model = "**vgg_ilsvrc_16_gender_imdb_wiki.onnx**"

Onnx files are very large (summary with caffe binaries - 3 Gb). I have uploaded the models to **google drive**.

**Comment by asiryan** [comment]
@wenbingl, why did the Linux CI and Windows CI checks fail?

**Review by matt-glich — APPROVED** [review]
finished on the overdue merged requests.

**Review by matt-glich — APPROVED** [review]

**Review by matt-glich — APPROVED** [review]

**Review by matt-glich — APPROVED** [review]

**Review by matt-glich — CHANGES_REQUESTED** [review]
binaries were not visible - references at the end describing "image manipulation".

**Comment by wenbingl** [comment]
> @wenbingl, why did the Linux CI and Windows CI checks fail?

I just rerun the job and it looks some random failure in the CI system? Any concern of checking in your model files into this repo by git-lfs?

**Comment by asiryan** [comment]
@wenbingl , I can push onnx models to this repo with git-lfs, it is ain't problem :)
But I solved problem with original Levi's technique with caffe models. VGG age and gender models by Rothe et al. are very large, but not accurate. Give me some time, I will redo this PR.

**Comment by wenbingl** [comment]
> @wenbingl , I can push onnx models to this repo with git-lfs, it is ain't problem :)
> But I solved problem with original Levi's technique with caffe models. VGG age and gender models by Rothe et al. are very large, but not accurate. Give me some time, I will redo this PR.

That's great, Thanks!

**Comment by asiryan** [comment]
@wenbingl, I pushed new commits, but they didn't pass Model-Zoo-Windows-CI validation. Could you check please what I am doing wrong?

**Comment by wenbingl** [comment]
> @wenbingl, I pushed new commits, but they didn't pass Model-Zoo-Windows-CI validation. Could you check please what I am doing wrong?

it looks CI test failure due to OOM, @jcwchen, any suggestion for that?

**Comment by jcwchen** [comment]
I think these models are good. Continuously loading and checking large ONNX models will cause memory issues in some machines. Currently the CI from Github Action is the same one as the AzurePipeline one. I would say we can deprecate Azure Pipeline now. @askhade What do you think?

**Comment by wenbingl** [comment]
> I think these models are good. Continuously loading and checking large ONNX models will cause memory issues in some machines. Currently the CI from Github Action is the same one as the AzurePipeline one. I would say we can deprecate Azure Pipeline now. @askhade What do you think?

Can you have a local buddy build to see if the CI process can run a large model on Windows with sufficient memory? If still fails, we may need some improvement, like split CI tasks into several Python files.

**Review by wenbingl — APPROVED** [review]

**Comment by jcwchen** [comment]
> > I think these models are good. Continuously loading and checking large ONNX models will cause memory issues in some machines. Currently the CI from Github Action is the same one as the AzurePipeline one. I would say we can deprecate Azure Pipeline now. @askhade What do you think?
> 
> Can you have a local buddy build to see if the CI process can run a large model on Windows with sufficient memory? If still fails, we may need some improvement, like split CI tasks into several Python files.
> 
> BTW, I don't have enough privilege to do a force merge, so the pipeline issue has to be addressed before it can be merged.

Yes actually on my local Windows continuously loading large models will encounter the MemoryError as well..., but on Mac-CI of GitHub Action won't hit the error so it happens on some environments due to the memory limitation. As your suggestion, I am trying to solve it by testing a model a time in script-level. I will update the CI here if I have something useful.

BTW, I will send a PR to deprecate AzurePipeline-CI soon. Thank you.
